lib/contextual_logger.rb

```diff
@@ -5,94 +5,98 @@
 require_relative './contextual_logger/context/handler'
 
 module ContextualLogger
-  def self.new(logger)
-    logger.extend(self)
+  class << self
+    def new(logger)
+      logger.extend(LoggerMixin)
+    end
   end
 
+  module LoggerMixin
   def global_context=(context)
-    ContextualLogger::Context::Handler.new(context).set!
+    Context::Handler.new(context).set!
   end
 
   def with_context(context)
-    context_handler = ContextualLogger::Context::Handler.new(current_context_for_thread.deep_merge(context))
+    context_handler = Context::Handler.new(current_context_for_thread.deep_merge(context))
     context_handler.set!
     if block_given?
       yield
     else
-      context_handler
+      context_handler # JEB, is this actually useful? The set! and reset! wouldn't mean anything, so it would just be calling Context::Handler.new, right? -Colin
```

Comment:
@jebentier

Reply:
This important because if a block is not given, we will be setting the new context, but then not be able to reset it when it's necessary to.  By using `with_context` without passing a block, it is returning the handler to the caller so that they can control when to reset the context.  For instance, if the context is already being set within a `begin/catch` block, and the user would like to not have the context reset before the exception is caught.  This is following a similar pattern seen in other gems that have the concept of temporary overrides, where you can follow the pattern of passing a block to execute for the duration of the temporary override, or you can get the reset handler back, and have control over when the override is removed.

Reply:
That's a really sneaky contract. I'll add a comment.

Reply:
I added a commend and also clarified the implementation, I think, by moving the `ensure` down to share the `if block_given?`.